### ISSUE TITLE
CVE-2024-41818 - fast-xml-parser

### ISSUE THREAD

**Issue by WtfJoke** [alexcasalboni/aws-lambda-power-tuning#263]
TLDR: Is it possible that you update your `@aws-sdk/*` version to >v3.621.0 (or latest)?

When deployed the latest version of aws-lambda-power-tuning we get a SecurityHub Alert: "CVE-2024-41818 - fast-xml-parser".

fast-xml-parser is a transitive dependency of the `@aws-sdk/*`. 
The vulnerable version of fast-xml-parser is not used anymore on `@aws-sdk/*` with versions >v3.621.0 according to https://github.com/aws/aws-sdk-js-v3/issues/6331#issuecomment-2259180119

**Comment by alexcasalboni** [comment]
Hi @WtfJoke the Lambda functions that are deployed should already use `@aws-sdk/client-lambda@3.651.1` and `@aws-sdk/client-s3@3.651.1`  (see [here](https://github.com/alexcasalboni/aws-lambda-power-tuning/blob/master/layer-sdk/package.json#L10)).

I guess the alert could be related to this repo's dev dependencies (who still use version `3.485.0` of the SDK). But that should be an easy fix :)

**Comment by alexcasalboni** [comment]
Fixed here: https://github.com/alexcasalboni/aws-lambda-power-tuning/commit/badfaa4ff39a3e6e92ce688ac8b84b812f3324bd

Unless you're using an older app version via SAR, you shouldn't get any more alerts. 

Could you please confirm? Please reopen this issue if you still get alarms :)

**Comment by WtfJoke** [comment]
@alexcasalboni Thanks for answering so fast :)  I hope you read this comment, I cannot reopen the issue. I was wondering why its a dev-dependency, thanks for confirming its a different package.json :D

We use the latest version from the SAR (4.3.5) since 2024-09-19 (e.g since 49 days as of today). It uses layer version 5 `arn:aws:lambda:eu-west-1:193133064756:layer:AWS-SDK-v3:5`

The scan finding is 24 days old (2024-10-14, sadly I dont know how I can retrigger a Security Hub Scan).

I have now following questions:
1. The changes from https://github.com/alexcasalboni/aws-lambda-power-tuning/commit/badfaa4ff39a3e6e92ce688ac8b84b812f3324bd needs a new SAR version/release right (to verify if this might have resulted in an alert)?
2. Is there some way to inspect the layer or how can we verify that this layer version contains the commit https://github.com/alexcasalboni/aws-lambda-power-tuning/commit/e56a7d98bb6b810734bdbbbc5fff3b15558ee966?

**Comment by WtfJoke** [comment]
> 2\. Is there some way to inspect the layer or how can we verify that this layer version contains the commit [e56a7d9](https://github.com/alexcasalboni/aws-lambda-power-tuning/commit/e56a7d98bb6b810734bdbbbc5fff3b15558ee966)?

I just redeployed the power tuner in a new aws account and it seems the layer uses the outdated aws-sdk version.
I inspected the layer by running this cli command:

`aws lambda get-layer-version-by-arn --arn arn:aws:lambda:eu-central-1:AWSACCOUNTID:layer:AWS-SDK-v3:1`

and downloaded the [AWS-SDK-v3-d4a82bb6-e40c-472d-bb03-292f02c2afed.zip](https://github.com/user-attachments/files/17659214/AWS-SDK-v3-d4a82bb6-e40c-472d-bb03-292f02c2afed.zip). The package.json in there still contains the aws-sdk version 3.485.0 with the vulnerable version of the fast-xml-parser 4.2.5 dependency.

Content:
```json
{
  "name": "aws-sdk-layer",
  "version": "1.0.0",
  "description": "A layer used by Lambda functions in this project.",
  "author": "Alex Casalboni <alex@alexcasalboni.com>",
  "licence": "MIT-0",
  "scripts": {},
  "private": true,
  "dependencies": {
    "@aws-sdk/client-lambda": "^3.485.0",
    "@aws-sdk/client-s3": "^3.485.0"
  }
}
```

TLDR: It seems that the newest lambda layer still uses the vulnerable version in their package.json and needs a fix.

**Comment by alexcasalboni** [comment]
@WtfJoke you're right! That's because I have not deployed a new version of the SAR app yet, so the fix is only available if you deploy "manually" by forking this repo.

> I was wondering why its a dev-dependency, thanks for confirming its a different package.json :D

The dev dependencies is only used for local development. Ideally it should be 100% aligned with the deployed SDK/Layer. I'll make sure to update both at the same time in the future. That said, it's odd that you get a warning about a dev dependency (which is not deployed anywhere since only the Lambda Layer is used in practice for all the deployed functions). I guess SecurityHub doesn't know the difference and just checks all dependencies.

> Is there some way to inspect the layer or how can we verify that this layer version contains the commit https://github.com/alexcasalboni/aws-lambda-power-tuning/commit/e56a7d98bb6b810734bdbbbc5fff3b15558ee966?

The best way to check what's included in a SAR version is checking the [ad-hoc SAR CHANGELOG](https://github.com/alexcasalboni/aws-lambda-power-tuning/blob/master/CHANGELOG.md).

I'm going to deploy a new SAR version later today, so you should be able to deploy a new stack with `v4.3.5`.

**Comment by WtfJoke** [comment]
Please check my comment above, there seems to be an issue with the deployment of the lambda layer...

**Comment by alexcasalboni** [comment]
What version of the SAR app have you deployed?

**Comment by WtfJoke** [comment]
I've deployed the newest version 4.3.5, I went to  https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:451282441545:applications~aws-lambda-power-tuning and pressed "Deploy" (I confirmed it, just again because its in the tag serverlessrepo:semanticVersion of the cloudformation stack as well).

**Comment by alexcasalboni** [comment]
Thanks for confirming!

Let me dig into this.

**Comment by WtfJoke** [comment]
I've updated the SAR app now to version 4.3.6 and the package.json of the [layer.zip](https://github.com/user-attachments/files/17660089/AWS-SDK-v3-7a6c9544-c6ff-4f26-80b7-56bb9417561a.zip) looks good now (does not contain the vulnerable dependency version).

```json
{
  "name": "aws-sdk-layer",
  "version": "1.0.0",
  "description": "A layer used by Lambda functions in this project.",
  "author": "Alex Casalboni <alex@alexcasalboni.com>",
  "licence": "MIT-0",
  "scripts": {},
  "private": true,
  "dependencies": {
    "@aws-sdk/client-lambda": "^3.651.1",
    "@aws-sdk/client-s3": "^3.651.1"
  }
}
```

Thanks for digging into this. 
Could this be a timing issue on the release of 4.3.5 (that it mistakenly was released before dependabot updated the package.json)?

**Comment by alexcasalboni** [comment]
@WtfJoke nice!

> Could this be a timing issue on the release of 4.3.5 (that it mistakenly was released before dependabot updated the package.json)?

That's possible!

This should be fixed both on  version `4.3.6` and `4.3.7` - although I'm having some trouble with verifying the actual ayer content (the package.json looks good though!).

If you encounter issues in the future with SAR (which is in "keeping the lights on" mode afaik), I highly recommend you start deploying Lambda Power Tuning directly with SAM:

```bash
$ sam build -u
$ sam deploy -g
```

**Comment by WtfJoke** [comment]
> although I'm having some trouble with verifying the actual layer content (the package.json looks good though!).

It contains the package.json and the package.lock.json, nothing else (I linked it above).




> If you encounter issues in the future with SAR (which is in "keeping the lights on" mode afaik), I highly recommend you start deploying Lambda Power Tuning directly with SAM:

Thanks for the hint! We deploy it using terraform with no issues so far (on each terraform deploy we deploy the latest version) :)

```hcl
data "aws_serverlessapplicationrepository_application" "lambda-power-tuning" {
  application_id = "arn:aws:serverlessrepo:us-east-1:451282441545:applications/aws-lambda-power-tuning"
}

resource "aws_serverlessapplicationrepository_cloudformation_stack" "lambda-power-tuning" {
  name           = "lambda-power-tuning"
  application_id = "arn:aws:serverlessrepo:us-east-1:451282441545:applications/aws-lambda-power-tuning"
  capabilities   = ["CAPABILITY_IAM"]
  # Always use latest version of the application
  semantic_version = data.aws_serverlessapplicationrepository_application.lambda-power-tuning.semantic_version
}
```

**Comment by alexcasalboni** [comment]
@WtfJoke nice! if you are using Terraform and want to drop SAR at some point, check out this on the official Terraform Registry as well: https://registry.terraform.io/modules/aws-ia/lambda-power-tuning/aws/latest
